modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py

```diff
@@ -137,6 +138,152 @@ def get_indices(cls, axis, partitions, index_func=None):
         new_idx = ray.get(new_idx)
         return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
 
+    @classmethod
+    def concat(cls, axis, left_parts, right_parts):
+        """
+        Concatenate the blocks of partitions with another set of blocks.
+
+        Parameters
+        ----------
+        axis : int
+            The axis to concatenate to.
+        left_parts : np.ndarray
+            NumPy array of partitions to concatenate with.
+        right_parts : np.ndarray or list
+            NumPy array of partitions to be concatenated.
+
+        Returns
+        -------
+        np.ndarray
+            A new NumPy array with concatenated partitions.
+
+        Notes
+        -----
+        Assumes that the blocks are already the same shape on the
+        dimension being concatenated. A ValueError will be thrown if this
+        condition is not met.
+        """
+        result = super(PandasOnRayDataframePartitionManager, cls).concat(
+            axis, left_parts, right_parts
+        )
+        if axis == 0:
+            return cls.rebalance_partitions(result)
+        else:
+            return result
+
+    @classmethod
+    def rebalance_partitions(cls, partitions):
+        """
+        Rebalance a 2-d array of partitions.
+
+        Rebalance the partitions by building a new array
+        of partitions out of the original ones so that:
+          - If all partitions have a length, each new partition has roughly the
+            same number of rows from the original partitions.
+          - Otherwise, each new partition spans roughly the same number of old
+            partitions.
+
+        Parameters
+        ----------
+        partitions : np.ndarray
+            The 2-d array of partitions to rebalance.
+
+        Returns
+        -------
+        np.ndarray
+            A new NumPy array with rebalanced partitions.
+        """
+        # We rebalance when the ratio of the number of existing partitions to
+        # the ideal number of partitions is larger than this threshold. The
+        # threshold is a heuristic that may need to be tuned for performance.
+        max_excess_of_num_partitions = 1.5
+        num_existing_partitions = partitions.shape[0]
+        ideal_num_new_partitions = NPartitions.get()
+        if (
+            num_existing_partitions
+            <= ideal_num_new_partitions * max_excess_of_num_partitions
+        ):
+            return partitions
+        # If any partition has an unknown length, give each axis partition
+        # roughly the same number of row partitions.
+        if any(partition.length() is None for row in partitions for partition in row):
+            # We need each partition to go into an axis partition, but the
+            # number of axis partitions may not evenly divide the number of
+            # partitions.
+            ideal_existing_partitions_per_new_partition = compute_chunksize(
+                num_existing_partitions, ideal_num_new_partitions
```

Comment:
A whole element count in an axis should be provided as the first argument. Why do we pass `num_existing_partitions`?

Reply:
Yes, it should be the whole length

Reply:
@mvashishtha why was this originally changed?